pkg/agent/controller/egress/egress_controller.go

```diff
@@ -505,6 +518,59 @@ func (c *EgressController) realizeEgressIP(egressName, egressIP string) (uint32,
 	return ipState.mark, nil
 }
 
+func trafficShapingToRateLimit(trafficShaping *crdv1b1.TrafficShaping) *RateLimit {
+	if trafficShaping == nil {
+		return nil
+	}
+	shapingStrToKbpsInt := func(s string) int {
+		num, _ := strconv.Atoi(s[:len(s)-1])
+		unit := s[len(s)-1:]
+		switch unit {
+		case "G":
+			num = num * 1000000
+		case "M":
+			num = num * 1000
+		}
+		return num
+	}
+	return &RateLimit{
+		Rate:  uint32(shapingStrToKbpsInt(trafficShaping.Rate)),
+		Burst: uint32(shapingStrToKbpsInt(trafficShaping.Burst)),
+	}
+}
+
+func (c *EgressController) realizeEgressQoS(eState *egressState, egressName string, mark uint32, trafficShaping *crdv1b1.TrafficShaping, exist bool) error {
+	klog.Infof("======= im in")
+	klog.Infof("egressname: %s mark: %s oldts: %s new ts: %s", egressName, mark, eState.rateLimit, trafficShaping)
+	if !c.trafficShapingEnabled {
+		if trafficShaping != nil {
+			klog.InfoS("TrafficShaping in the Egress is ignored because OVS meters are not supported or "+
+				"trafficShaping is not enabled in Antrea-agent config.", "EgressName", egressName)
+		}
+		return nil
+	}
+	curRateLimit := trafficShapingToRateLimit(trafficShaping)
+	qosNeedInstall := func(preRateLimit, curRateLimit *RateLimit, mark uint32) bool {
+		if curRateLimit != nil && mark != 0 && (preRateLimit == nil || curRateLimit.Rate != preRateLimit.Rate || curRateLimit.Burst != preRateLimit.Burst) {
+			return true
+		}
+		return false
+	}
+	if qosNeedInstall(eState.rateLimit, curRateLimit, mark) {
+		if err := c.installEgressQoS(mark, trafficShapingToRateLimit(trafficShaping)); err != nil {
+			return err
+		}
+		eState.rateLimit = curRateLimit
+	}
+	if (curRateLimit == nil || mark == 0) && eState.rateLimit != nil && exist {
+		if err := c.uninstallEgressQoS(eState.mark); err != nil {
+			return err
+		}
+		eState.rateLimit = nil
+	}
+	return nil
```

Comment:
If a non-local IP becomes local then the mark will be changed. We can't skip only checking `eState.rateLimit.Equals(curRateLimit)`. As for the uninstall, for example, if users only changed the bandwidth config, the mark doesn't change, then `eState.rateLimit != nil` will delete what we just installed.

Reply:
> If a non-local IP becomes local then the mark will be changed. We can't skip only checking `eState.rateLimit.Equals(curRateLimit)`.

`eState.rateLimit` is nil if a non-local IP becomes local, right? We should only store rateLimit in eState when we have configured it in datapath. But there is indeed an issue in the proposed code, we should check mark too when determing the desired rateLimit.
```
	var desiredRateLimit *rateLimit
        // QoS is desired only if the Egress is configured on this Node.
        if mark != nil {
            desiredRateLimit = trafficShapingToRateLimit(trafficShaping)
        }
	// Nothing changes.
	if eState.rateLimit.Equals(desiredRateLimit) {
		return nil
	}
	// It's desired to have QoS on this Node, install/override it.
	if desiredRateLimit != nil {
		if err := c.installEgressQoS(mark, desiredRateLimit); err != nil {
			return err
		}
		eState.rateLimit = desiredRateLimit
		return nil
	}
	// It's undesired to have QoS on this Node, uninstall it.
	if eState.rateLimit != nil {
		if err := c.uninstallEgressQoS(eState.mark); err != nil {
			return err
		}
		eState.rateLimit = nil
	}
```

> As for the uninstall, for example, if users only changed the bandwidth config, the mark doesn't change, then `eState.rateLimit != nil` will delete what we just installed.

No, if `desiredRateLimit` is not nil, it will only install QoS and return early.

The proposed code handles each case and returns early.
If nothing changes, it returns early. Then there must be a change in the following code.
If the desired rate limit is not empty, it installs the desired one and returns early.
If it comes to the last check, there is no desired rate limit, it uninstall previous rate limit if there is one.

Reply:
Thanks for the explanation. You are right. Changed.